src/ember-app.js

```diff
@@ -287,77 +111,203 @@ class EmberApp {
-   * information from its `package.json` file.
-   */
-  readPackageJSON(distPath) {
-    let pkgPath = path.join(distPath, 'package.json');
-    let file;
-
-    try {
-      file = fs.readFileSync(pkgPath);
-    } catch (e) {
-      throw new Error(`Couldn't find ${pkgPath}. You may need to update your version of ember-cli-fastboot.`);
+/**
+ * Initializes the sandbox by evaluating the Ember app's JavaScript
+ * code, then retrieves the application factory from the sandbox and creates a new
+ * `Ember.Application`.
+ *
+ * @returns {Ember.Application} the Ember application from the sandbox
+ */
+function createSandboxedEmberApp(sandbox, appFilePath, vendorFilePath) {
+
+  loadAppFiles(sandbox, appFilePath, vendorFilePath);
+
+  // Retrieve the application factory from within the sandbox
+  let AppFactory = sandbox.run(function(ctx) {
+    return ctx.require('~fastboot/app-factory');
+  });
+
+  // If the application factory couldn't be found, throw an error
+  if (!AppFactory || typeof AppFactory['default'] !== 'function') {
+    throw new Error('Failed to load Ember app from ' + appFilePath + ', make sure it was built for FastBoot with the `ember fastboot:build` command.');
+  }
+
+  // Otherwise, return a new `Ember.Application` instance
+  return AppFactory['default']();
+}
+
+/**
+ *
+ * Loads the app and vendor files in the sandbox (Node vm).
+ *
+ */
+function loadAppFiles(sandbox, appFilePath, vendorFilePath) {
+  sandbox.eval('sourceMapSupport.install(Error);');
+
+  let appFile = fs.readFileSync(appFilePath, 'utf8');
+  let vendorFile = fs.readFileSync(vendorFilePath, 'utf8');
+
+  debug("evaluating app; app=%s; vendor=%s", appFilePath, vendorFilePath);
+
+  sandbox.eval(vendorFile, vendorFilePath);
+  debug("vendor file evaluated");
+
+  sandbox.eval(appFile, appFilePath);
+  debug("app file evaluated");
+}
+
+/**
+ * Creates a destroy function to destroy the created `ApplicationInstance`.
+ *
+ * @param {Object} The Result object that contains the application instance to be destroyed.
+ * @param {Integer} How long to wait before manually destroying the application instance. Ignored if its value is 0.
+ *
+ * @returns {Function} The destructor function.
+ */
+function destroyAppInstance(instance, result, timeout) {
+  if (!timeout) return () => {
+    debug('destroying application instance');
+    instance.destroy();
+  };
+
+  // start a timer to destroy the appInstance forcefully in the given ms.
+  // This is a failure mechanism so that node process doesn't get wedged if the `visit` never completes.
+  let destructionTimer = setTimeout(function() {
+    debug('destroying application instance because of timeout');
+    instance.destroy();
+    result.error = new Error(`App instance was forcefully destroyed in ${timeout}ms`);
+  }, timeout);
+
+  return () => {
+    debug('destroying application instance');
+    clearTimeout(destructionTimer);
+    instance.destroy();
```

Comment:
What if instance was destroyed earlier due to the timeout? This will result in an error I think.